man/man1/tabcols.1
.TH TABCOLS 1NEMO "20 June 2026"

.SH "NAME"
tabcols \- Select columns, and optionally rearrange, from a table

.SH "SYNOPSIS"
\fPtabcols\fP [parameter=value] ...

.SH "DESCRIPTION"
\fBtabcols\fP selects columns from a file, and can thus also re-arrange the
output column order. New column separators can be used on output.
This way a newline on input can be converted into a space on 
output and therefore creating a file with one long column from
a table with multiple columns.
.PP
To select specific rows/lines, use \fItabrows(1NEMO)\fP or \fItabmath(1NEMO)\fP.
.PP
To align a table by column, use the \fIcolumn(1)\fP program with the \fB-t\fP
option, or use the \fBwidth=\fP parameter

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP
input file name(s).
.br
No default.
.TP
\fBselect=\fP
list of columns to select. Any \fInemoinp(1NEMO)\fP  expression can be used.
Columns can be repeated, and in any order. The output will be in exactly
the same order as given with this keyword. First column is 1. Column 0 can also
be added to the output, which is the row number (first row is 1).
.br
[all]    
.TP
\fBcolsep=\fP
Column separator. Valid are: \fBs\fP (space), \fBn\fP (newline), 
\fBt\fP (tab), 
\fBr\fP (carriage return), as well as the literal symbols '\fB,\fP' 
and '\fB:\fP'.
.br
[Default: \fBs\fP]

.TP
\fBcolsepin=\fP
Column separator for input. By default an internal set (space, comma, tab, vertical bar)
is used. A string can be specified here consistent of multiple characters.

.TP
\fBempty=t|f\fP
Allow empty columns? For example "1,,3" would then return an empty string. Use with care,
since column separators also matter.
.br
Default: f

.TP
\fBwidth=\fP
Specify the width of the extracted columns. If fewer numbers than the actual number,
the width value repeats.
.br
Default: no specified widths
.TP
\fBout=-\fP
output file name. By default standard output it used.
.br
[-]

.SH "CAVEATS"
Although this program uses the new table interface, there are
still some internal variables that limit the column count.

.SH "EXAMPLE"
Here is an example of how to reverse the columns from a table with 4 columns:
.nf
    tabcols in.tab 4:1:-1 > out.tab
.fi
and here is an example that
re-arranges all the numbers in a table in one long column
.nf
    tabcols in.tab all n > out.tab
.fi
or in one long row:
.nf
    tabcols in.tab all n | tabtranspose - - > out.tab
.fi

.SH "SEE ALSO"
tabrows(1NEMO), tabtranspos(1NEMO), tabmath(1NEMO), tabs(1NEMO), tabtab(1NEMO), awk(1), column(1)

.SH "FILES"
src/kernel/tab/tabcols.c   source code

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.5i +5.0i
26-Jan-00	V1.0 Created 	PJT
9-apr-09	V2.0 out=- now default and last argument	PJT
5-may-2022	V2.1 converted to table V2 interface	PJT
10-sep-2024	V2.4 added colsepin=	PJT
20-jun-2026	V3.1 added width= and make empty=f default	PJT
.fi


src/kernel/tab/tabcols.c
/*
 * TABCOLS: select columns from a table; a simple awk like processor
 *
 *      27-jan-00   created
 *
 *   @todo     properly parse ,, as an empty column (use burst0string)
 */

#include <stdinc.h>
#include <getparam.h>
#include <table.h>
#include <extstring.h>
#include <ctype.h>

string defv[] = {                /* DEFAULT INPUT PARAMETERS */
    "in=???\n           input file name(s)",
    "select=all\n       columns to select (add 0 to add column with row number)",
    "colsep=SP\n        Column separator for output (SP,TAB,NL)",
    "colsepin=\n        Optional enforced character to separate columns in input",
    "empty=t\n          Allow empty strings (only useful of colsep is not SP)",
    "width=\n           Enforce a width for each output column (repeats if absent)",
    "out=-\n            output file name",
    "VERSION=3.1\n      11-jun-2026 PJT",
    NULL
};

string usage = "Select column(s) from a table";


string input, output;			/* file names */
stream instr, outstr;			/* file streams */
table  *tptr;                           /* table */
int    ninput;				/* number of input files */

#ifndef MAX_COL
#define MAX_COL 256
#endif


int    keep[MAX_COL+1];                 /* column numbers to keep */
int    width[MAX_COL-1];                /* width of columns, if set */
int    nkeep, nwidth;                   /* actual number of skip columns */
int    maxcol = 0;                      /* largest column number */
string colsepin;                        /* character(s) to separate columns on input */
char   colsep;                          /* character to separate columns on output */
bool   Qall;
bool   Qempty;
char   fmt[32];

local void setparams(void);
local void convert(stream , stream);
local void tab2space(char *);

extern  string *burststring(string, string);
extern  string *burst0string(string, string);


void nemo_main()
{
    setparams();
    convert (instr,outstr);
}

local void setparams(void)
{
    string select;                          /* which columns not to write */
    string separator;
    int i;

    input = getparam("in");
    output = getparam("out");

    instr  = stropen(input,"r");
    tptr   = table_open(instr,1);
    outstr = stropen (output,"w");

    Qempty = getbparam("empty");

    nwidth = nemoinpi(getparam("width"),width,MAX_COL);    

    // dprintf(1,"table: %d x %d\n", table_nrows(tptr), table_ncols(tptr));
    

    select = getparam("select");
    if (select==NULL || *select=='\0' || streq(select,"all")) {
        Qall = TRUE;
        for (i=0; i<MAX_COL; i++) keep[i] = i+1;
    } else {
        Qall = FALSE;
        nkeep = nemoinpi(select,keep,MAX_COL);
        if (nkeep<=0 || nkeep>MAX_COL)
            error("Too many columns given (%d) to keep (MAX_COL %d)",nkeep,MAX_COL);
	// @todo allow -1 as the last columns
	if (nkeep==1 && keep[0]==-1) {
	  warning("new feature -1");
	} else {
	  for (i=0; i<nkeep; i++){
            if (keep[i]<0) error("Negative column number %d",keep[i]);
            maxcol = MAX(maxcol,keep[i]);
	  }
        }
    }
    dprintf(1,"nkeep::%d\n",nkeep);
    dprintf(1,"nwidth::%d\n",nwidth);
    if (nwidth > 0) dprintf(1,"width::%d\n",width[0]);
    if (nkeep > 0) {
      for (i=nwidth; i<nkeep; i++)
	width[i] = width[i-1];
    }

    separator = getparam("colsep");
    switch (*separator) {
        case 's':
        case 'S':       colsep =  ' ';   break;
        case 't':
        case 'T':       colsep = '\t';  break;
        case 'n':
        case 'N':       colsep = '\n';  break;
        case 'r':
        case 'R':       colsep = '\n';  break;
        case ':':       
        case ',':       
        case '|':
        case '\\':
        case '/':       colsep = separator[0]; break;
        default:        error("Illegal separator: s t n r : ,");
    }
    dprintf(1,"colsep:'%c'\n", colsep);

    colsepin = getparam("colsepin");
    dprintf(1,"colsepin:'%s'\n", colsepin);
}


local void convert(stream instr, stream outstr)
{
    char   *line;
    int    i, nlines, noutv;
    string *outv;                   /* pointer to vector of strings to write */
    char   *seps;

    if (strlen(colsepin) == 0)
      seps = strdup(", |\t");
    else
      seps = strdup(colsepin);
    
    nlines=0;               /* count lines read so far */

    for (;;) {

        line = table_line(tptr);
	if (line == NULL) return;

        dprintf(3,"LINE: (%s)\n",line);
        if (iscomment(line)) continue;
        
        nlines++;
        tab2space(line);                  /* work around a Gipsy (?) problem */

	if (Qempty)
	  outv = burst0string(line,seps);
	else
	  outv = burststring(line,seps);
	  
        noutv = xstrlen(outv,sizeof(string)) - 1;
        if (noutv < maxcol) {
	  warning("skipping line %d; Too few columns in input file (%d < %d)",nlines,noutv,maxcol);
	  dprintf(2,"Line: %s\n",line);
	  continue;
	} else
	  dprintf(2,"Line[%d,%d] %s\n",nlines,noutv,line);
        if (Qall) nkeep = noutv;
                        
        for (i=0; i<nkeep; i++) {
            if (keep[i] == 0)
                fprintf(outstr,"%d",nlines);
            else {
	        if (nwidth > 0) {
		    sprintf(fmt,"%%-%ds",width[i]);
                    fprintf(outstr,fmt,outv[keep[i]-1]);
	        } else
                    fprintf(outstr,"%s",outv[keep[i]-1]);
	    }
            if (i < nkeep-1) fprintf(outstr,"%c",colsep);
        }
        if (colsep != 'n') fprintf(outstr,"\n");    /* end of line */

	// @todo    free
    }
}
/*
 * small helper function, replaces tabs by spaces before processing.
 * this prevents me from diving into gipsy parsing routines and fix
 * the problem there 
 * PJT - June 1998.
 */

local void tab2space(char *cp)
{
    while (*cp) {
        if (*cp == '\t') *cp = ' ';
        cp++;
    }
}
